pygmt/src/grdhisteq.py

```diff
@@ -0,0 +1,341 @@
+        shading of this grid (using :meth:`pygmt.Figure.grdimage` or
+        :meth:`pygmt.Figure.grdview`) with a linear mapping from topography to
+        graytone will result in most of the image being very dark gray, with
+        the mountain being almost white. :meth:`pygmt.grdhisteq` can provide a
+        list of data values that divide the data range into divisions which
+        have an equal area in the image [Default is 16 if ``divisions`` is not
+        set]. The :class:`pandas.DataFrame` or ASCII file output can be used to
+        make a colormap with :meth:`pygmt.makecpt` and an image with
+        :meth:`pygmt.Figure.grdimage` that has all levels of gray occuring
+        equally.
+
+        :meth:`pygmt.grdhisteq` also provides a way to write a grid with
+        statistics based on a cumulative distribution function. In this
+        application, the ``outgrid`` has relative highs and lows in the same
+        (x,y) locations as the ``grid``, but the values are changed to reflect
+        their place in the cumulative distribution.
+
+        Must provide ``outfile`` or ``outgrid``.
+
+        Full option list at :gmt-docs:`grdhisteq.html`
+
+        {aliases}
+
+        Parameters
+        ----------
+        grid : str or xarray.DataArray
+            The file name of the input grid or the grid loaded as a DataArray.
+        outgrid : str or bool or None
+            The name of the output netCDF file with extension .nc to store the
+            grid in.
+        outfile : str or bool or None
+            The name of the output ASCII file to store the results of the
+            histogram equalization in. Not allowed if ``outgrid`` is used.
+        divisions : int
+            Set the number of divisions of the data range [Default is 16].
+
+        {R}
+        {V}
+
+        Returns
+        -------
+        ret: pandas.DataFrame or xarray.DataArray or None
+            Return type depends on whether the ``outgrid`` parameter is set:
+
+            - pandas.DataFrame if ``outfile`` is True
+            - xarray.DataArray if ``outgrid`` is True
+            - None if ``outgrid`` is a str (grid output is stored in
+              ``outgrid``)
+            - None if ``outfile`` is a str (file output is stored in
+              ``outfile``)
+
+        See Also
+        -------
+        :meth:`pygmt.grd2cpt`
+        """
+
+        with Session() as lib:
+            file_context = lib.virtualfile_from_data(check_kind="raster", data=grid)
+            with file_context as infile:
+                arg_str = " ".join([infile, build_arg_string(kwargs)])
+                lib.call_module("grdhisteq", arg_str)
+
+        if output_type == "file":
+            return None
+        if output_type == "xarray":
+            return load_dataarray(tmpfile.name)
+
+        result = pd.read_csv(
+            kwargs["D"], sep="\t", header=None, names=["start", "stop", "bin_id"]
```

Comment:
Reduce memory footprint of output `pandas.DataFrame`. I suppose uint8/uint32 should work for most cases?

```suggestion
            filepath_or_buffer=kwargs["D"],
            sep="\t",
            header=None,
            names=["start", "stop", "bin_id"],
            dtype={
                "start": np.float32,
                "stop": np.float32,
                "bin_id": np.uint8 if kwargs["C"] < 256 else np.uint32,
            },
```

Reply:
This would require more logic because `kwargs["C"]` can be `None`. What do you think about just having np.uint32 for "bin_id"? This is most compatible with its representation in GMT and also simpler than having it depend on the number of divisions.